File: python/bingo_board.py
import string


class BingoBoard:

    def __init__(self, width: int, height: int):
        self.__cells = [ [None]*height for _ in range(width) ]
        self.__marked = [ [False]*height for _ in range(width) ]

    def define_cell(self, x: int, y: int, value: string):
        if self.__cells[x][y] is not None:
            raise IllegalStateException("Cell already defined.")

        for c in range(len(self.__cells)):
            for r in range(len(self.__cells[c])):
                if value == self.__cells[c][r]:
                    raise IllegalStateException(f"{value} already present at {c}, {r}")

        self.__cells[x][y] = value

    def mark_cell(self, x: int, y: int):
        if not self.is_initialised():
            raise IllegalStateException("Board is not initialised.")

        self.__marked[x][y] = True

    def is_initialised(self):
        for row in self.__cells:
            for column in row:
                if column is None:
                    return False

        return True

    def is_marked(self, x: int, y: int):
        return self.__marked[x][y]


class IllegalStateException(Exception):
    pass

File: go/bingo_board.go
package bingo_board

import (
	"errors"
	"fmt"
)

type BingoBoard struct {
	Cells  [][]*string
	Marked [][]bool
}

func New(width, height int) BingoBoard {
	cells := make([][]*string, width)
	for row := range width {
		cells[row] = make([]*string, height)
	}

	marked := make([][]bool, width)
	for row := range width {
		marked[row] = make([]bool, height)
	}

	return BingoBoard{
		Cells:  cells,
		Marked: marked,
	}
}

func (b *BingoBoard) DefineCell(x, y int, value string) error {
	if b.Cells[x][y] != nil {
		return errors.New("cell already defined")
	}

	for c := range b.Cells {
		for r := range b.Cells[c] {
			if b.Cells[c][r] != nil && value == *b.Cells[c][r] {
				return fmt.Errorf("%s already present at %v, %v", value, c, r)
			}
		}
	}

	b.Cells[x][y] = &value
	return nil
}

func (b *BingoBoard) MarkCell(x, y int) error {
	if !b.IsInitialised() {
		return errors.New("board is not initialised")
	}

	b.Marked[x][y] = true
	return nil
}

func (b *BingoBoard) IsInitialised() bool {
	for _, row := range b.Cells {
		for _, column := range row {
			if column == nil {
				return false
			}
		}
	}
	return true
}

func (b *BingoBoard) IsMarked(x, y int) bool {
	return b.Marked[x][y]
}
